crates/bevy_ecs/src/query/iter.rs

```diff
@@ -71,6 +72,136 @@ where
     }
 }
 
+/// An [`Iterator`] over query results of a [`Query`](crate::system::Query).
+///
+/// This struct is created by the [`Query::many_iter`](crate::system::Query::many_iter) and
+/// [`Query::many_iter_mut`](crate::system::Query::many_iter_mut) methods.
+pub struct QueryManyIter<
+    'w,
+    's,
+    Q: WorldQuery,
+    QF: Fetch<'w, State = Q::State>,
+    F: WorldQuery,
+    E: Borrow<Entity>,
+    I: Iterator<Item = E>,
+> {
+    entity_iter: I,
+    world: &'w World,
+    fetch: QF,
+    filter: QueryFetch<'w, F>,
+    query_state: &'s QueryState<Q, F>,
+}
+
+impl<
+        'w,
+        's,
+        Q: WorldQuery,
+        QF: Fetch<'w, State = Q::State>,
+        F: WorldQuery,
+        E: Borrow<Entity>,
+        I: Iterator<Item = E>,
+    > QueryManyIter<'w, 's, Q, QF, F, E, I>
+{
+    /// # Safety
+    /// This does not check for mutable query correctness. To be safe, make sure mutable queries
+    /// have unique access to the components they query.
+    /// This does not validate that `world.id()` matches `query_state.world_id`. Calling this on a `world`
+    /// with a mismatched [`WorldId`](crate::world::WorldId) is unsound.
+    pub(crate) unsafe fn new<EntityList: IntoIterator<IntoIter = I>>(
+        world: &'w World,
+        query_state: &'s QueryState<Q, F>,
+        entity_list: EntityList,
+        last_change_tick: u32,
+        change_tick: u32,
+    ) -> QueryManyIter<'w, 's, Q, QF, F, E, I> {
+        let fetch = QF::init(
+            world,
+            &query_state.fetch_state,
+            last_change_tick,
+            change_tick,
+        );
+        let filter = QueryFetch::<F>::init(
+            world,
+            &query_state.filter_state,
+            last_change_tick,
+            change_tick,
+        );
+        QueryManyIter {
+            query_state,
+            world,
+            fetch,
+            filter,
+            entity_iter: entity_list.into_iter(),
+        }
+    }
+}
+
+impl<
+        'w,
+        's,
+        Q: WorldQuery,
+        QF: Fetch<'w, State = Q::State>,
+        F: WorldQuery,
+        E: Borrow<Entity>,
+        I: Iterator<Item = E>,
+    > Iterator for QueryManyIter<'w, 'w, Q, QF, F, E, I>
+{
+    type Item = QF::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            let World {
+                entities,
+                archetypes,
+                storages,
```

Comment:
Can you use a `&'w` reference to  all three of these instead of in `next`? No need to derefence a `&'w World` in a hot loop when you can avoid it.